include/sparrow/layout/array_registry.hpp

```diff
@@ -0,0 +1,442 @@
+         *
+         * @tparam F Visitor function type (must be callable with all array types)
+         * @param func The visitor function to apply
+         * @param ar The array wrapper to visit
+         * @return The result of invoking func with the concrete array type
+         * @throws std::invalid_argument if array type is not supported
+         * @throws std::runtime_error if dictionary data type is not an integer
+         *
+         * @example
+         * // Works with both base types and extensions
+         * auto& registry = array_registry::instance();
+         *
+         * // Register a custom extension
+         * registry.register_extension(data_type::BINARY, "my.extension", my_factory);
+         *
+         * // Create and dispatch - extensions are automatically detected
+         * auto wrapper = registry.create(some_proxy);
+         * auto size = registry.dispatch([](auto&& arr) { return arr.size(); }, *wrapper);
+         */
+        template <class F>
+        [[nodiscard]] visit_result_t<F> dispatch(F&& func, const array_wrapper& ar) const;
+
+    private:
+
+        array_registry() = default;
+
+        // Helper for dispatching with compile-time type knowledge
+        template <class F, data_type DT>
+        static auto dispatch_for_type(F&& func, const array_wrapper& ar) -> visit_result_t<F>
+        {
+            if constexpr (DT == data_type::TIMESTAMP_SECONDS || DT == data_type::TIMESTAMP_MILLISECONDS
+                          || DT == data_type::TIMESTAMP_MICROSECONDS || DT == data_type::TIMESTAMP_NANOSECONDS)
+            {
+                // Special handling for timestamp types with timezone check
+                using types = timestamp_type_map<DT>;
+                if (get_timezone(ar.get_arrow_proxy()) == nullptr)
+                {
+                    return func(unwrap_array<typename types::without_tz>(ar));
+                }
+                else
+                {
+                    return func(unwrap_array<typename types::with_tz>(ar));
+                }
+            }
+            else
+            {
+                return func(unwrap_array<array_type_t<DT>>(ar));
+            }
+        }
+
+        // Recursive helper for dispatching to the correct type
+        template <class F, std::size_t I = 0>
+        static auto try_dispatch_recursive(F&& func, const array_wrapper& ar, data_type dt)
+            -> visit_result_t<F>
+        {
+            if constexpr (I < all_data_types.size())
+            {
+                if (all_data_types[I] == dt)
+                {
+                    return dispatch_for_type<F, all_data_types[I]>(std::forward<F>(func), ar);
+                }
+                else
+                {
+                    return try_dispatch_recursive<F, I + 1>(std::forward<F>(func), ar, dt);
+                }
+            }
+            else
+            {
+                throw std::invalid_argument("array type not supported");
+            }
```

Comment:
I think we could get rid of the recursion with a "table", something like

```cpp
template <class F>
struct invoker
{
    template <class I>
    static auto run(F& f, const array_wrapper& ar)
    {
        return f(unwrap_array<array_type_t<I>>(ar));
    }
};

template <class F, std::size_t... I>
constexpr auto make_invoker_table(F& f, std::index_sequence<I...>)
{
    return std::array<invoker_type, all_data_types.size()>{&invoker<F>::run<I>...};
}

template <class F>
visit_result_t<F> dispatch(F&& f, const array_wrapper& ar)
{
    const auto dt = ar.data_type();
    if (dt >= all_data_types.size())
    {
        throw ...;
    }
    return make_invoker_table(f, std::make_index_sequence(all_data_types.size())[dt](ar);
}
```

This also avoids checking boundary condition on dt at each step.